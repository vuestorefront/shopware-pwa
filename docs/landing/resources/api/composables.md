<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [@shopware-pwa/composables](./composables.md)

## composables package

## Functions

|  Function | Description |
|  --- | --- |
|  [createListingComposable({ rootContext, searchMethod, searchDefaults, listingKey, })](./composables.createlistingcomposable.md) | <b><i>(BETA)</i></b> Factory to create your own listing. By default you can use useListing composable, which provides you predefined listings for category(cms) listing and product search listing. Using factory you can provide our own compatible search method and use it for example for creating listing of orders in my account. |
|  [getApplicationContext(rootContext, key)](./composables.getapplicationcontext.md) | <b><i>(BETA)</i></b> |
|  [getDefaultApiParams()](./composables.getdefaultapiparams.md) | <b><i>(BETA)</i></b> Returns default system API params |
|  [useBreadcrumbs(rootContext, params)](./composables.usebreadcrumbs.md) | <b><i>(BETA)</i></b> Composable for displaying and setting breadcrumbs for page. |
|  [useCms(rootContext)](./composables.usecms.md) | <b><i>(BETA)</i></b> |
|  [useProductAssociations(rootContext, product, association)](./composables.useproductassociations.md) | <b><i>(BETA)</i></b> Get product association entity. Options - [IUseProductAssociations](./composables.iuseproductassociations.md) |
|  [useSharedState(rootContext)](./composables.usesharedstate.md) | <b><i>(BETA)</i></b> Replacement for Vuex. Composable, which enables you to use shared state in your application. State is shared both on server and client side. |

## Interfaces

|  Interface | Description |
|  --- | --- |
|  [IInterceptorCallbackFunction](./composables.iinterceptorcallbackfunction.md) | <b><i>(BETA)</i></b> interface for the callback function of interceptors |
|  [IUseAddToCart](./composables.iuseaddtocart.md) | <b><i>(BETA)</i></b> interface for [useAddToCart](./composables.useaddtocart.md) composable |
|  [IUseCart](./composables.iusecart.md) | <b><i>(BETA)</i></b> interface for [useCart](./composables.usecart.md) composable |
|  [IUseCheckout](./composables.iusecheckout.md) | <b><i>(BETA)</i></b> interface for [useCheckout](./composables.usecheckout.md) composable |
|  [IUseIntercept](./composables.iuseintercept.md) | <b><i>(BETA)</i></b> interface for [useIntercept](./composables.useintercept.md) composable |
|  [IUseListing](./composables.iuselisting.md) | <b><i>(BETA)</i></b> Listing interface, can be used to display category products, search products or any other Shopware search interface (ex. orders with pagination) |
|  [IUseNavigation](./composables.iusenavigation.md) | <b><i>(BETA)</i></b> interface for [useNavigation](./composables.usenavigation.md) composable<!-- -->Provides state for navigation trees depending on navigation type. |
|  [IUseProductAssociations](./composables.iuseproductassociations.md) | <b><i>(BETA)</i></b> interface for [IUseProductAssociations](./composables.iuseproductassociations.md) composable |
|  [IUseProductConfigurator](./composables.iuseproductconfigurator.md) | <b><i>(BETA)</i></b> interface for [useProductConfigurator](./composables.useproductconfigurator.md) composable |
|  [IUseProductQuickSearch](./composables.iuseproductquicksearch.md) | <b><i>(BETA)</i></b> |
|  [IUseSessionContext](./composables.iusesessioncontext.md) | <b><i>(BETA)</i></b> interface for [useSessionContext](./composables.usesessioncontext.md) composable |
|  [IUseUser](./composables.iuseuser.md) | <b><i>(BETA)</i></b> interface for [useUser](./composables.useuser.md) composable |
|  [IUseWishlist](./composables.iusewishlist.md) | <b><i>(BETA)</i></b> interface for [useWishlist](./composables.usewishlist.md) composable |
|  [Notification\_2](./composables.notification_2.md) | <b><i>(BETA)</i></b> |
|  [Routing](./composables.routing.md) | <b><i>(BETA)</i></b> |
|  [UseCountries](./composables.usecountries.md) | <b><i>(BETA)</i></b> |
|  [UseCountry](./composables.usecountry.md) | <b><i>(BETA)</i></b> |
|  [UseCurrency](./composables.usecurrency.md) | <b><i>(BETA)</i></b> |
|  [UseProduct](./composables.useproduct.md) | <b><i>(BETA)</i></b> |
|  [UseSalutations](./composables.usesalutations.md) | <b><i>(BETA)</i></b> |

## Variables

|  Variable | Description |
|  --- | --- |
|  [INTERCEPTOR\_KEYS](./composables.interceptor_keys.md) | <b><i>(BETA)</i></b> Keys used accross composables with the description of incommint parameters. |
|  [useAddToCart](./composables.useaddtocart.md) | <b><i>(BETA)</i></b> Add product to cart. Options - [IUseAddToCart](./composables.iuseaddtocart.md) |
|  [useCart](./composables.usecart.md) | <b><i>(BETA)</i></b> Composable for cart management. Options - [IUseCart](./composables.iusecart.md) |
|  [useCategoryFilters](./composables.usecategoryfilters.md) | <b><i>(BETA)</i></b> |
|  [useCheckout](./composables.usecheckout.md) | <b><i>(BETA)</i></b> Composable for Checkout management. Options - [IUseCheckout](./composables.iusecheckout.md) |
|  [useCountries](./composables.usecountries.md) | <b><i>(BETA)</i></b> |
|  [useCountry](./composables.usecountry.md) | <b><i>(BETA)</i></b> |
|  [useCurrency](./composables.usecurrency.md) | <b><i>(BETA)</i></b> |
|  [useDefaults](./composables.usedefaults.md) | <b><i>(BETA)</i></b> Returns default config depending on config key. It is used in composables, so defaultsKey is in most cases composable name (ex. <code>useDefaults(rootContext, &quot;useCms&quot;)</code>) |
|  [useIntercept](./composables.useintercept.md) | <b><i>(BETA)</i></b> Allows to broadcast and intercept events across application. |
|  [useListing](./composables.uselisting.md) | <b><i>(BETA)</i></b> |
|  [useNavigation](./composables.usenavigation.md) | <b><i>(BETA)</i></b> Composable for navigation. Options - [IUseNavigation](./composables.iusenavigation.md) |
|  [useNotifications](./composables.usenotifications.md) | <b><i>(BETA)</i></b> |
|  [useProduct](./composables.useproduct.md) | <b><i>(BETA)</i></b> |
|  [useProductConfigurator](./composables.useproductconfigurator.md) | <b><i>(BETA)</i></b> Product options - [IUseAddToCart](./composables.iuseaddtocart.md) |
|  [useProductQuickSearch](./composables.useproductquicksearch.md) | <b><i>(BETA)</i></b> |
|  [useSalutations](./composables.usesalutations.md) | <b><i>(BETA)</i></b> |
|  [useSessionContext](./composables.usesessioncontext.md) | <b><i>(BETA)</i></b> Composable for session management. Options - [IUseSessionContext](./composables.iusesessioncontext.md) |
|  [useUIState](./composables.useuistate.md) | <b><i>(BETA)</i></b> Simple state management for UI purposes. |
|  [useUser](./composables.useuser.md) | <b><i>(BETA)</i></b> Composable for user management. Options - [IUseUser](./composables.iuseuser.md) |
|  [useWishlist](./composables.usewishlist.md) | <b><i>(BETA)</i></b> |

## Type Aliases

|  Type Alias | Description |
|  --- | --- |
|  [ApplicationVueContext](./composables.applicationvuecontext.md) | <b><i>(BETA)</i></b> Application Context for Shopware PWA. It's an extended Vue instance. |
|  [listingKey](./composables.listingkey.md) | <b><i>(BETA)</i></b> |
|  [Search](./composables.search.md) | <b><i>(BETA)</i></b> |

