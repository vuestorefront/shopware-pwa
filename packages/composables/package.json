{
  "name": "@shopware-pwa/composables",
  "version": "1.0.1",
  "description": "@shopware-pwa/composables",
  "main": "dist/composables.cjs.js",
  "module": "dist/composables.esm.js",
  "files": [
    "dist"
  ],
  "types": "dist/composables.d.ts",
  "unpkg": "dist/composables.global.js",
  "sideEffects": false,
  "repository": {
    "type": "git",
    "url": "git+https://github.com/DivanteLtd/shopware-pwa"
  },
  "buildOptions": {
    "name": "composables",
    "formats": [
      "esm",
      "cjs"
    ]
  },
  "license": "MIT",
  "dependencies": {
    "@shopware-pwa/helpers": "1.0.1",
    "@shopware-pwa/shopware-6-client": "1.0.1",
    "@vue/devtools-api": "^6.0.0-beta.19",
    "axios": "^0.24.0",
    "cookie-universal": "^2.1.5",
    "lodash": "^4.17.21",
    "query-string": "^7.0.1",
    "vue-demi": "^0.12.0"
  },
  "peerDependencies": {
    "@vue/composition-api": "^1.3.1",
    "vue": "^2.0.0 || >=3.0.0-rc.0"
  },
  "peerDependenciesMeta": {
    "@vue/composition-api": {
      "optional": true
    }
  },
  "devDependencies": {
    "@vue/composition-api": "^1.3.1",
    "vue": "^2.6.14"
  },
  "publishConfig": {
    "access": "public"
  }
}
