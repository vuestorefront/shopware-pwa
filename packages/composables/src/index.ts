/**
 * Vue's composables to be used in Shopware frontend application.
 *
 * @packageDocumentation
 */

import { ApiDefaults } from "@shopware-pwa/commons";
import { ShopwareApiInstance } from "@shopware-pwa/shopware-6-client";
import {
  App,
  markRaw,
  effectScope,
  EffectScope,
  reactive,
  isVue2,
} from "vue-demi";
import { registerShopwareDevtools } from "./devtools/plugin";

export * from "./hooks/useCms";
export * from "./hooks/useProduct";
export * from "./hooks/useCart";
export * from "./logic/useAddToCart";
export * from "./logic/useCheckout";
export * from "./logic/useSessionContext";
export * from "./logic/useUIState";
export * from "./hooks/useCurrency";
export * from "./hooks/useNavigation";
export * from "./hooks/useSalutations";
export * from "./hooks/useCountries";
export * from "./hooks/useUser";
export * from "./hooks/useCustomerOrders";
export * from "./hooks/useCustomerPassword";
export * from "./hooks/useCustomerAddresses";
export * from "./logic/useDefaults";
export * from "./logic/useNotifications";
export * from "./logic/useIntercept";
export * from "./getDefaultApiParams";
export * from "./logic/useListing";
export * from "./logic/useProductQuickSearch";
export * from "./factories/createListingComposable";
export * from "./logic/useWishlist";
export * from "./hooks/useCountry";
export * from "./logic/useProductConfigurator";
export * from "./logic/useBreadcrumbs";
export * from "./logic/useProductAssociations";
export * from "./logic/useSharedState";
export * from "./logic/useOrderDetails";
export * from "./getVueContext";
export {
  getApplicationContext,
  ShopwareDomain,
  SwRouting,
  SwInterceptors,
  SwInterceptor,
} from "./appContext";

export { ShopwareVuePlugin } from "./devtools/vue2";

/* istanbul ignore next */
/**
 * Create ShopwarePlugin vue instance. Shpware PWA composables rely on this config.
 *
 * @beta
 */
export function createShopware(
  app: App,
  options: {
    initialStore: any;
    shopwareDefaults: ApiDefaults;
    apiInstance: ShopwareApiInstance;
  }
) {
  const scope: EffectScope = effectScope(true);
  const state = scope.run(() => {
    return reactive({
      interceptors: {},
      sharedStore: options.initialStore || reactive({}),
      shopwareDefaults: options.shopwareDefaults || {},
    });
  });

  const shopwarePlugin = markRaw({
    install(
      app: App,
      options?: {
        enableDevtools: boolean;
      }
    ) {
      if (!isVue2) {
        shopwarePlugin._a = app;
        (app as any).config.globalProperties.$shopware = shopwarePlugin;
        (app as any).provide("shopware", shopwarePlugin);
        /* istanbul ignore else */
        if (options?.enableDevtools && typeof window !== "undefined") {
          registerShopwareDevtools(app, shopwarePlugin);
        }
      }
    },
    _a: app,
    _e: scope,
    apiInstance: options.apiInstance,
    state,
  });

  return shopwarePlugin;
}
