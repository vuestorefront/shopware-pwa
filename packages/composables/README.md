# Shopware PWA composables

> [!WARNING]
> Deprecated. This package is deprecated. Look out for the new project [Showpare Composable Frontends](https://frontends.shopware.com/).
>
> Direct replacement of this package is [@shopware/composables](https://npmjs.com/package/@shopware/composables) package

Vue 3 composables dedicated for Shopware.  
Package implements Vue Storefront Next architecture and provides agnostic composables for eCommerce logic.

## Installing

Using yarn:

```bash
$ yarn add @shopware-pwa/composables
```

## Usage examples

this example uses `@shopware-pwa/shopware-6-client` package for loading product data

```js
import { useProduct } from "@shopware-pwa/composables"

// in component
setup (props) {
  const productId = "hiufh3847fhb4y3uvf"
  const { product, search } = useProduct()
  search(productId)

  return { product }
},
```
