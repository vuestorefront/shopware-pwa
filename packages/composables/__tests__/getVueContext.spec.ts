import { extendScopeContext, useVueContext } from "@shopware-pwa/composables";
import vueComp from "vue-demi";
const mockedCompositionAPI = vueComp as jest.Mocked<typeof vueComp>;

describe("Shopware composables - getVueContext", () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  describe("useVueContext", () => {
    it("should return vue context indicators", () => {
      const context = useVueContext();
      expect(context).toEqual({ isVueComponent: false, isVueScope: false });
    });

    it("should indicate vue component", () => {
      mockedCompositionAPI.getCurrentInstance = jest
        .fn()
        .mockReturnValue({ proxy: {} });
      const { isVueComponent } = useVueContext();
      expect(isVueComponent).toBe(true);
    });
    it("should indicate vue scope", () => {
      mockedCompositionAPI.getCurrentScope = jest
        .fn()
        .mockReturnValue({ vm: {} });
      const { isVueScope } = useVueContext();
      expect(isVueScope).toBe(true);
    });
  });

  describe("extendScopeContext", () => {
    it("should extend scope context", () => {
      const result = {
        vm: {},
      };
      extendScopeContext(result, {});
      expect(result).toMatchSnapshot();
    });
  });
});
