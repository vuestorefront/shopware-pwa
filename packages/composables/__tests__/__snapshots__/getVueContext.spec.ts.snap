// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shopware composables - getVueContext extendScopeContext should extend scope context 1`] = `
Object {
  "vm": Object {
    "i18n": undefined,
    "interceptors": undefined,
    "routing": undefined,
    "sharedStore": undefined,
    "shopware": undefined,
    "shopwareApiInstance": undefined,
    "shopwareDefaults": undefined,
  },
}
`;
